contracts/AddressSet.sol

```diff
@@ -2,40 +2,139 @@ pragma solidity ^0.4.18;
 
 
 library AddressSet {
+
+    //--- Definitions
+
+    struct Link {
+        address previous;
+        address next;
+    }
+
     struct Data {
-        address[] list;
-        mapping (address => uint256) index; // 1..n, 0 means it does not exists
+        address head;
+        address tail;
+        uint256 count;
+        mapping (address => Link) links;
     }
 
-    function add(Data storage self, address element) public {
-        if (self.index[element] == 0) {
-            self.list.push(element);
-            self.index[element] = self.list.length;
+    //--- Public mutable functions
+
+    function add(
+        Data storage self,
+        address element
+    )
+        internal
+        returns (bool added)
+    {
+        require(element != address(0));
+
+        if (self.count == 0) {
+            self.head = element;
+            added = true;
+        } else if (!contains(self, element)) {
+            self.links[self.tail].next = element;
+            self.links[element].previous = self.tail;
+            added = true;
+        }
+
+        if (added) {
+            self.tail = element;
+            safeIncrement(self);
         }
     }
 
-    function remove(Data storage self, address element) public {
-        // preserves initial order of elements, can be optimized if not required
+    function remove(
+        Data storage self,
+        address element
+    )
+        internal
+        returns (bool removed)
+    {
+        require(element != address(0));
+
+        Link storage link = self.links[element];
+
```

Comment:
I think this code can be made simpler by:
- checking if element exists in the beginning and returning false if not.
- Then switching the order of `head` / `tail` checking if's, like this and removing the `else if` condition:
```
        if (element == self.head) {
            self.head = link.next;
        } else {
            self.links[link.previous].next = link.next;
        }  
```
- removing the `if (removed) part`, it's safeguarded in the beginning
- and then always explicitly returning true in the end.

Reply:
I changed it a little differently than you suggested. You can call it micro-optimization, but I believe that low gas usage is a very important in widely used contracts. However, I tried to keep it as clean as possible.